etcdserver/server.go

```diff
@@ -570,12 +576,53 @@ type etcdProgress struct {
 	appliedi  uint64
 }
 
+// readyHandler contains a set of EtcdServer operations to be called by raftNode,
+// and helps decouple state machine logic from Raft algorithms.
+// TODO: add a state machine interface to apply the commit entries and do snapshot/recover
+type readyHandler struct {
+	leadershipUpdate func(lead bool) <-chan time.Time
+	sync             func()
+	sendMessage      func(msgs []raftpb.Message)
+}
+
 func (s *EtcdServer) run() {
 	snap, err := s.r.raftStorage.Snapshot()
 	if err != nil {
 		plog.Panicf("get snapshot from raft storage error: %v", err)
 	}
-	s.r.start(s)
+
+	rh := &readyHandler{
+		leadershipUpdate: func(lead bool) (sync <-chan time.Time) {
+			switch lead {
```

Comment:
the switch is sort of weird since it's only true/false, how about:
```go
if !lead {
    if s.lessor != nil { s.lessor.Demote() }
    if s.compactor != nil { s.compactor.Pause() }
    return nil
}
if s.stats != nil { s.stats.BecomeLeader() }
if s.comapctor != nil { s.compactor.Resume() }
if s.td != nil { s.td.Reset() }
return s.SyncTicker
```